course02/homework01/Tomasz/list.h

```diff
@@ -18,41 +19,64 @@ class List
 	};
 
 public:
-	template<typename E>
-	class Iterator
+	class ConstIterator
 	{
-	private:
-		Iterator(const List<T>& owner, Node* node);
+	protected:
+		ConstIterator(const List<T>& owner, Node* node);
 
 	public:
-		Iterator operator++(int);
-		Iterator& operator++();
-		Iterator operator--(int);
-		Iterator& operator--();
+		using value_type = T;
+		using difference_type = std::ptrdiff_t;
+		using reference = T & ;
+		using pointer = T * ;
+		using iterator_category = std::bidirectional_iterator_tag;
 
-		E& operator*() const;
-		E* operator->() const;
+		ConstIterator operator++(int);
+		ConstIterator& operator++();
+		ConstIterator operator--(int);
+		ConstIterator& operator--();
 
-		bool operator==(const Iterator& second);
-		bool operator!=(const Iterator& second);
+		const T& operator*() const;
+		const T* operator->() const;
 
-	private:
-		const List<T>& mOwner;
+		bool operator==(const ConstIterator& second) const;
+		bool operator!=(const ConstIterator& second) const;
+
+	protected:
+		const List<T>* mOwner;
 		Node* mNode;
 
 		friend List<T>;
 	};
 
-	using value_type = T;
+	class MutableIterator : public ConstIterator
```

Comment:
The pattern used in general to implement iterators is to have an `IteratorBase` templated class, and to inherit from it from Iterator and ConstIterator